include/powerloader/target.hpp

```diff
@@ -41,47 +41,182 @@ namespace powerloader
+        // Requirement: `prepare_for_transfer()` must have been called successfully before calling
+        // this function.
+        void set_to_max_speed();
+
+        void reset();
+        void reset_response();
+
+        void prepare_for_transfer(CURLM* multi_handle,
+                                  const std::string& full_url,
+                                  Protocol protocol);
+
+        void flush_target_file();
+
+        tl::expected<void, DownloaderError> finish_transfer(const std::string& effective_url);
+
+        void complete_mirror_usage(bool was_success,
+                                   const tl::expected<void, DownloaderError>& result);
+
+        bool can_retry_transfer_with_fewer_connections() const;
+        void lower_mirror_parallel_connections();
+
+        const DownloadTarget& target() const noexcept
+        {
+            assert(m_target);
+            return *m_target;
+        }
+
+        // TODO: don't allow external code from manipulating this internal target.
+        DownloadTarget& target() noexcept
+        {
+            assert(m_target);
+            return *m_target;
+        }
+
+        // TODO: don't expose ownership
+        const std::shared_ptr<Mirror>& mirror() const noexcept
+        {
+            assert(m_mirror);
+            return m_mirror;
+        }
+
+        // TODO: don't expose ownership
+        std::shared_ptr<Mirror>& mirror() noexcept
+        {
+            assert(m_mirror);
+            return m_mirror;
+        }
+
+        void change_mirror(std::shared_ptr<Mirror> mirror);
+
+        const fs::path temp_file() const noexcept
+        {
+            return m_temp_file;
+        }
+
+        bool writecb_required_range_written() const noexcept
+        {
+            return m_writecb_required_range_written;
+        }
+
+        HeaderCbState headercb_state() const noexcept
+        {
+            return m_headercb_state;
+        }
+
+        const std::string& headercb_interrupt_reason() const noexcept
+        {
+            return m_headercb_interrupt_reason;
+        }
+
+        int range_fail() const noexcept
+        {
+            return m_range_fail;
+        }
+
+        void reset_range_fail() noexcept
+        {
+            m_range_fail = false;
+        }
+
+        std::size_t retries() const noexcept
+        {
+            return m_retries;
+        }
+
+        DownloadState state() const noexcept
+        {
+            return m_state;
+        }
+
+        ZckState zck_state() const noexcept
+        {
+            return m_zck_state;
+        }
+
+        // TODO: refactor to avoid state being changed directly from outisde.
+        void set_zck_state(ZckState new_state) noexcept
+        {
+            m_zck_state = new_state;
+        }
+
+        const auto& errorbuffer() const noexcept
+        {
+            return m_errorbuffer;
+        }
+
+        const auto& mirrors() const noexcept
+        {
+            return m_mirrors;
```

Comment:
With a short alias for `m_mirrors`'s type, the return type of the method could be more explicit.

Reply:
Ok noted for next PR